flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -147,6 +150,148 @@ static void processFirstRow(
+	 * @param out                   underlying collector
+	 * @param generateUpdateBefore  flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert        flag to gennerate INSERT message or not
+	 * @param orderFunctionProvider provider that provides an order function to judge first or last
+	 */
+	public static void deduplicateOnRowTime(
+			ValueState<RowData> state,
+			RowData currentRow,
+			Collector<RowData> out,
+			boolean generateUpdateBefore,
+			boolean generateInsert,
+			DeduplicateOrderFunctionProvider orderFunctionProvider) throws Exception {
+
+		checkInsertOnly(currentRow);
+		RowData preRow = state.value();
+		if (!orderFunctionProvider.matches(preRow, currentRow)) {
+			return;
+		}
+
+		collectRetractResult(
+				generateUpdateBefore,
+				generateInsert,
+				preRow,
+				currentRow,
+				out,
+				null
+		);
+		state.update(currentRow);
+	}
+
+	/**
+	 * Processes element to deduplicate on keys with row time semantic, sends current element if it is last
+	 * or first row, retracts previous element if needed.
+	 *
+	 * @param state                 state of function
+	 * @param bufferedRows          latest row received by deduplicate function
+	 * @param serializer            serializer to serialize the data
+	 * @param out                   underlying collector
+	 * @param generateUpdateBefore  flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert        flag to gennerate INSERT message or not
+	 * @param orderFunctionProvider provider that provides an order function to judge first or last
+	 */
+	public static void miniBatchDeduplicateOnRowTime(
+			ValueState<RowData> state,
+			List<RowData> bufferedRows,
+			TypeSerializer<RowData> serializer,
+			Collector<RowData> out,
+			boolean generateUpdateBefore,
+			boolean generateInsert,
+			DeduplicateOrderFunctionProvider orderFunctionProvider) throws Exception {
+		if (bufferedRows == null) {
+			return;
+		}
+
+		RowData preRow = state.value();
+		for (RowData currentRow : bufferedRows) {
+			checkInsertOnly(currentRow);
+			if (!orderFunctionProvider.matches(preRow, currentRow)) {
+				continue;
+			}
+			collectRetractResult(
+					generateUpdateBefore,
+					generateInsert,
+					preRow,
+					currentRow,
+					out,
+					serializer
+			);
+			preRow = currentRow;
+		}
+		state.update(preRow);
+	}
+
+	private static void collectRetractResult(
+			boolean generateUpdateBefore,
+			boolean generateInsert,
+			RowData preRow,
+			RowData currentRow,
+			Collector<RowData> out,
+			TypeSerializer<RowData> serializer) {
+		if (generateUpdateBefore || generateInsert) {
+			if (preRow == null) {
+				// the first row, send INSERT message
+				currentRow.setRowKind(RowKind.INSERT);
+				out.collect(currentRow);
+			} else {
+				if (generateUpdateBefore) {
+					RowData copyRow;
+					// when miniBatch enabled, do a copy here, the serializer is not null
+					if (serializer != null) {
+						copyRow = serializer.copy(preRow);
```

Comment:
Why we have to do a deep copy here?
If it is for not affecting the original `preRow`, we can set back `RowKind` after collecting `preRow` (non-minibatch mode also has this problem).

Reply:
If this copying can be removed, the implementations of `collectRetractResult` and `processLastRowOnProcTime` are totally the same. I feel that `collectRetractResult` is not suitable, because it not always generates retractions. 

What about combine them in the the following new method?

```java
static void updateDeduplicateResult(
			RowData currentRow,  // new row result of deduplicate under current key
			ValueState<RowData> state, // the state stores the previous collected row under current key
			boolean generateUpdateBefore,           
			boolean generateInsert,
			Collector<RowData> out) throws Exception {
```

Reply:
the state update logic is a little different, I skip this one